File: typescript/design_pattern/bridge/main_bridge.js
"use strict";
/**
 * Bridge Design Pattern - TypeScript Implementation
 *
 * The Bridge pattern separates an abstraction from its implementation
 * so that both can vary independently. This example demonstrates
 * remote controls (abstraction) operating different devices (implementation).
 */
// =============================================================================
// Concrete Implementations - TV and Radio
// =============================================================================
/**
 * TV - A concrete implementation of the Device interface
 */
class TV {
    constructor() {
        this.enabled = false;
        this.volume = 30;
        this.channel = 1;
    }
    isEnabled() {
        return this.enabled;
    }
    enable() {
        this.enabled = true;
        console.log("[Line 42] TV: Powered ON");
    }
    disable() {
        this.enabled = false;
        console.log("[Line 47] TV: Powered OFF");
    }
    getVolume() {
        return this.volume;
    }
    setVolume(volume) {
        if (volume > 100) {
            this.volume = 100;
        }
        else if (volume < 0) {
            this.volume = 0;
        }
        else {
            this.volume = volume;
        }
        console.log(`[Line 62] TV: Volume set to ${this.volume}`);
    }
    getChannel() {
        return this.channel;
    }
    setChannel(channel) {
        this.channel = channel;
        console.log(`[Line 70] TV: Channel set to ${this.channel}`);
    }
    getName() {
        return "TV";
    }
}
/**
 * Radio - Another concrete implementation of the Device interface
 */
class Radio {
    constructor() {
        this.enabled = false;
        this.volume = 20;
        this.channel = 88;
    }
    isEnabled() {
        return this.enabled;
    }
    enable() {
        this.enabled = true;
        console.log("[Line 92] Radio: Powered ON");
    }
    disable() {
        this.enabled = false;
        console.log("[Line 97] Radio: Powered OFF");
    }
    getVolume() {
        return this.volume;
    }
    setVolume(volume) {
        if (volume > 100) {
            this.volume = 100;
        }
        else if (volume < 0) {
            this.volume = 0;
        }
        else {
            this.volume = volume;
        }
        console.log(`[Line 112] Radio: Volume set to ${this.volume}`);
    }
    getChannel() {
        return this.channel;
    }
    setChannel(channel) {
        this.channel = channel;
        console.log(`[Line 120] Radio: Station set to ${this.channel} FM`);
    }
    getName() {
        return "Radio";
    }
}
// =============================================================================
// Abstraction - RemoteControl
// =============================================================================
/**
 * The Abstraction defines the interface for the "control" part of the two
 * class hierarchies. It maintains a reference to an object of the
 * Implementation hierarchy and delegates all real work to this object.
 */
class RemoteControl {
    constructor(device) {
        this.device = device;
        console.log(`[Line 142] RemoteControl: Created for ${device.getName()}`);
    }
    togglePower() {
        console.log(`[Line 146] RemoteControl: Toggle power button pressed`);
        if (this.device.isEnabled()) {
            this.device.disable();
        }
        else {
            this.device.enable();
        }
    }
    volumeDown() {
        console.log(`[Line 155] RemoteControl: Volume down button pressed`);
        this.device.setVolume(this.device.getVolume() - 10);
    }
    volumeUp() {
        console.log(`[Line 160] RemoteControl: Volume up button pressed`);
        this.device.setVolume(this.device.getVolume() + 10);
    }
    channelDown() {
        console.log(`[Line 165] RemoteControl: Channel down button pressed`);
        this.device.setChannel(this.device.getChannel() - 1);
    }
    channelUp() {
        console.log(`[Line 170] RemoteControl: Channel up button pressed`);
        this.device.setChannel(this.device.getChannel() + 1);
    }
    getStatus() {
        console.log(`[Line 175] RemoteControl: Status - ${this.device.getName()} is ${this.device.isEnabled() ? "ON" : "OFF"}, Volume: ${this.device.getVolume()}, Channel: ${this.device.getChannel()}`);
    }
}
// =============================================================================
// Refined Abstraction - AdvancedRemoteControl
// =============================================================================
/**
 * AdvancedRemoteControl extends the basic RemoteControl with additional features.
 * This demonstrates that abstractions can be extended without affecting implementations.
 */
class AdvancedRemoteControl extends RemoteControl {
    constructor(device) {
        super(device);
        console.log(`[Line 191] AdvancedRemoteControl: Enhanced features enabled`);
    }
    mute() {
        console.log(`[Line 195] AdvancedRemoteControl: Mute button pressed`);
        this.device.setVolume(0);
    }
    setVolumeLevel(level) {
        console.log(`[Line 200] AdvancedRemoteControl: Setting exact volume to ${level}`);
        this.device.setVolume(level);
    }
    jumpToChannel(channel) {
        console.log(`[Line 205] AdvancedRemoteControl: Jumping to channel ${channel}`);
        this.device.setChannel(channel);
    }
    printDetailedStatus() {
        const status = this.device.isEnabled() ? "ON" : "OFF";
        console.log(`[Line 211] AdvancedRemoteControl: Detailed Status Report`);
        console.log(`[Line 212]   - Device: ${this.device.getName()}`);
        console.log(`[Line 213]   - Power: ${status}`);
        console.log(`[Line 214]   - Volume: ${this.device.getVolume()}%`);
        console.log(`[Line 215]   - Channel: ${this.device.getChannel()}`);
    }
}
// =============================================================================
// Client Code - Demonstration
// =============================================================================
function clientCode() {
    console.log("=".repeat(60));
    console.log("[Line 223] Bridge Pattern Demonstration");
    console.log("=".repeat(60));
    console.log();
    // Scenario 1: Basic Remote with TV
    console.log("[Line 227] --- Scenario 1: Basic Remote with TV ---");
    const tv = new TV();
    const tvRemote = new RemoteControl(tv);
    tvRemote.togglePower();
    tvRemote.volumeUp();
    tvRemote.volumeUp();
    tvRemote.channelUp();
    tvRemote.getStatus();
    tvRemote.togglePower();
    console.log();
    // Scenario 2: Basic Remote with Radio
    console.log("[Line 239] --- Scenario 2: Basic Remote with Radio ---");
    const radio = new Radio();
    const radioRemote = new RemoteControl(radio);
    radioRemote.togglePower();
    radioRemote.volumeUp();
    radioRemote.channelUp();
    radioRemote.getStatus();
    radioRemote.togglePower();
    console.log();
    // Scenario 3: Advanced Remote with TV
    console.log("[Line 250] --- Scenario 3: Advanced Remote with TV ---");
    const tv2 = new TV();
    const advancedTvRemote = new AdvancedRemoteControl(tv2);
    advancedTvRemote.togglePower();
    advancedTvRemote.setVolumeLevel(75);
    advancedTvRemote.jumpToChannel(42);
    advancedTvRemote.printDetailedStatus();
    advancedTvRemote.mute();
    advancedTvRemote.printDetailedStatus();
    console.log();
    // Scenario 4: Advanced Remote with Radio
    console.log("[Line 262] --- Scenario 4: Advanced Remote with Radio ---");
    const radio2 = new Radio();
    const advancedRadioRemote = new AdvancedRemoteControl(radio2);
    advancedRadioRemote.togglePower();
    advancedRadioRemote.setVolumeLevel(50);
    advancedRadioRemote.jumpToChannel(101);
    advancedRadioRemote.printDetailedStatus();
    console.log();
    console.log("=".repeat(60));
    console.log("[Line 272] Bridge Pattern Demonstration Complete");
    console.log("=".repeat(60));
}
// Run the demonstration
clientCode();

File: rust/design_pattern/bridge/main_bridge.rs
// Bridge Design Pattern in Rust
// Separates abstraction from implementation so both can vary independently

use std::cell::RefCell;

// Implementation trait - defines the interface for implementation classes
trait Device {
    fn is_enabled(&self) -> bool;
    fn enable(&mut self);
    fn disable(&mut self);
    fn get_volume(&self) -> u32;
    fn set_volume(&mut self, volume: u32);
    fn get_channel(&self) -> u32;
    fn set_channel(&mut self, channel: u32);
    fn get_name(&self) -> &str;
}

// Concrete Implementation A - TV
struct TV {
    enabled: bool,
    volume: u32,
    channel: u32,
}

impl TV {
    fn new() -> Self {
        println!("[TV] Creating new TV device");
        TV {
            enabled: false,
            volume: 30,
            channel: 1,
        }
    }
}

impl Device for TV {
    fn is_enabled(&self) -> bool {
        self.enabled
    }

    fn enable(&mut self) {
        self.enabled = true;
        println!("[TV] Device enabled");
    }

    fn disable(&mut self) {
        self.enabled = false;
        println!("[TV] Device disabled");
    }

    fn get_volume(&self) -> u32 {
        self.volume
    }

    fn set_volume(&mut self, volume: u32) {
        self.volume = volume.min(100);
        println!("[TV] Volume set to {}", self.volume);
    }

    fn get_channel(&self) -> u32 {
        self.channel
    }

    fn set_channel(&mut self, channel: u32) {
        self.channel = channel;
        println!("[TV] Channel set to {}", self.channel);
    }

    fn get_name(&self) -> &str {
        "TV"
    }
}

// Concrete Implementation B - Radio
struct Radio {
    enabled: bool,
    volume: u32,
    channel: u32,
}

impl Radio {
    fn new() -> Self {
        println!("[Radio] Creating new Radio device");
        Radio {
            enabled: false,
            volume: 20,
            channel: 88,
        }
    }
}

impl Device for Radio {
    fn is_enabled(&self) -> bool {
        self.enabled
    }

    fn enable(&mut self) {
        self.enabled = true;
        println!("[Radio] Device enabled");
    }

    fn disable(&mut self) {
        self.enabled = false;
        println!("[Radio] Device disabled");
    }

    fn get_volume(&self) -> u32 {
        self.volume
    }

    fn set_volume(&mut self, volume: u32) {
        self.volume = volume.min(100);
        println!("[Radio] Volume set to {}", self.volume);
    }

    fn get_channel(&self) -> u32 {
        self.channel
    }

    fn set_channel(&mut self, channel: u32) {
        self.channel = channel;
        println!("[Radio] Station set to {} FM", self.channel);
    }

    fn get_name(&self) -> &str {
        "Radio"
    }
}

// Abstraction - holds a reference to the implementation
struct RemoteControl {
    device: RefCell<Box<dyn Device>>,
}

impl RemoteControl {
    fn new(device: Box<dyn Device>) -> Self {
        println!("Creating RemoteControl for {}", device.get_name());
        RemoteControl {
            device: RefCell::new(device),
        }
    }

    fn toggle_power(&self) {
        let mut device = self.device.borrow_mut();
        if device.is_enabled() {
            device.disable();
        } else {
            device.enable();
        }
    }

    fn volume_up(&self) {
        let mut device = self.device.borrow_mut();
        let current = device.get_volume();
        device.set_volume(current + 10);
    }

    fn volume_down(&self) {
        let mut device = self.device.borrow_mut();
        let current = device.get_volume();
        if current >= 10 {
            device.set_volume(current - 10);
        } else {
            device.set_volume(0);
        }
    }

    fn channel_up(&self) {
        let mut device = self.device.borrow_mut();
        let current = device.get_channel();
        device.set_channel(current + 1);
    }

    fn channel_down(&self) {
        let mut device = self.device.borrow_mut();
        let current = device.get_channel();
        if current > 1 {
            device.set_channel(current - 1);
        }
    }

    fn get_device_name(&self) -> String {
        self.device.borrow().get_name().to_string()
    }
}

// Refined Abstraction - extends the base abstraction with additional features
struct AdvancedRemoteControl {
    base: RemoteControl,
}

impl AdvancedRemoteControl {
    fn new(device: Box<dyn Device>) -> Self {
        println!("Creating AdvancedRemoteControl for {}", device.get_name());
        AdvancedRemoteControl {
            base: RemoteControl {
                device: RefCell::new(device),
            },
        }
    }

    fn toggle_power(&self) {
        self.base.toggle_power();
    }

    fn volume_up(&self) {
        self.base.volume_up();
    }

    fn volume_down(&self) {
        self.base.volume_down();
    }

    fn mute(&self) {
        let mut device = self.base.device.borrow_mut();
        println!("[{}] Muting device", device.get_name());
        device.set_volume(0);
    }

    fn set_channel_direct(&self, channel: u32) {
        let mut device = self.base.device.borrow_mut();
        println!("[{}] Setting channel directly to {}", device.get_name(), channel);
        device.set_channel(channel);
    }

    fn get_device_name(&self) -> String {
        self.base.get_device_name()
    }
}

fn main() {
    println!("=== Bridge Design Pattern Demo ===\n");

    // Demo 1: Basic RemoteControl with TV
    println!("--- Demo 1: Basic RemoteControl with TV ---");
    let tv = Box::new(TV::new());
    let tv_remote = RemoteControl::new(tv);

    println!("\nOperating {} with basic remote:", tv_remote.get_device_name());
    tv_remote.toggle_power();
    tv_remote.volume_up();
    tv_remote.volume_up();
    tv_remote.channel_up();
    tv_remote.toggle_power();

    // Demo 2: Basic RemoteControl with Radio
    println!("\n--- Demo 2: Basic RemoteControl with Radio ---");
    let radio = Box::new(Radio::new());
    let radio_remote = RemoteControl::new(radio);

    println!("\nOperating {} with basic remote:", radio_remote.get_device_name());
    radio_remote.toggle_power();
    radio_remote.volume_up();
    radio_remote.channel_up();
    radio_remote.channel_up();

    // Demo 3: AdvancedRemoteControl with TV
    println!("\n--- Demo 3: AdvancedRemoteControl with TV ---");
    let tv2 = Box::new(TV::new());
    let advanced_remote = AdvancedRemoteControl::new(tv2);

    println!("\nOperating {} with advanced remote:", advanced_remote.get_device_name());
    advanced_remote.toggle_power();
    advanced_remote.volume_up();
    advanced_remote.set_channel_direct(42);
    advanced_remote.mute();
    advanced_remote.toggle_power();

    // Demo 4: AdvancedRemoteControl with Radio
    println!("\n--- Demo 4: AdvancedRemoteControl with Radio ---");
    let radio2 = Box::new(Radio::new());
    let advanced_radio_remote = AdvancedRemoteControl::new(radio2);

    println!("\nOperating {} with advanced remote:", advanced_radio_remote.get_device_name());
    advanced_radio_remote.toggle_power();
    advanced_radio_remote.set_channel_direct(101);
    advanced_radio_remote.volume_up();
    advanced_radio_remote.mute();

    println!("\n=== Bridge Pattern Demo Complete ===");
}
